### PR TITLE
Further editing for dynamic allocation best practices

### PR DESCRIPTION
**Pull request by camilbancioiu**

### CODE REVIEW

**Review by dorin-iancu — APPROVED** [review]

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 1]
```diff
@@ -266,44 +266,68 @@ The first approach looks a lot simpler, just a `ManagedVec` of tuples. But, the
...
 | Unmanaged | Managed |
 | :---: | :---: |
-| &[u8] | TokenIdentifier |
-| - | BigUint |
-| BoxedBytes | ManagedBuffer |
-| String | ManagedBuffer |
-| VarArgs | ManagedVarArgs |
-| Vec | ManagedVec |
-| MultiResultVec | ManagedMultiResultVec |
+| `&[u8]` | `TokenIdentifier` |
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:304]: A little bit misleading .... &[u8] corresponds to &ManagedBuffer
There is no unmanaged TokenIdentifier left ... that one was (painfully) upgraded without leaving anything behind

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 2]
```diff
@@ -266,44 +266,68 @@ The first approach looks a lot simpler, just a `ManagedVec` of tuples. But, the
...
+| - | `BigUint` |
+| `BoxedBytes` | `ManagedBuffer` |
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:306]: Also add `Vec<u8>` alongside BoxedBytes .... same row or another ....

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 3]
```diff
@@ -266,44 +266,68 @@ The first approach looks a lot simpler, just a `ManagedVec` of tuples. But, the
...
+You can use the `erdpy contract report` command to verify whether your contract still requires dynamic allocation or not.
+:::
 
 ### Static buffers
 
-Sometimes you might want to allocate some memory inside the SC, because it might be easier to use. In that case, we recommend using static buffers. The declaration looks like this:
+Sometimes you might want to allocate some memory inside the smart contract, because it might be easier to use. In that case, we recommend using static buffers. The declaration looks like this:
 
 ```rust
 static mut STATIC_BUFFER: [u8; BUFFER_SIZE] = [0u8; BUFFER_SIZE];
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:327]: We should **NOT** encourage this, it is not thread-safe for the tests.
We need to write a section on how to use the StaticBufferRef type, which references the big static buffer.

Conversely, it is safe to temporarily use arrays allocated on the stack.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 4]
```diff
@@ -266,44 +266,68 @@ The first approach looks a lot simpler, just a `ManagedVec` of tuples. But, the
...
+
+In most cases, the managed types can be used as drop-in replacements for the basic Rust types. For a simple example, see [BigUint Operations](#biguint-operations).
+
+:::warning
+Make sure you migrate to the managed types **incrementally** and **thoroughly test your code** before even considering deploying to the mainnet.
+:::
+
+:::tip
+You can use the `erdpy contract report` command to verify whether your contract still requires dynamic allocation or not.
+:::
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:321]: Let's also add a section about ArrayVec, which is easy to use, safe & fully integrated in elrond-wasm.
Only downside is that you need a max cap.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 5]
```diff
@@ -198,16 +198,16 @@ Keep in mind that all the mappers can have as many additional arguments for the
...
+The Rust language does not natively support var args. The best you could probably do in native rust is use `Option<T>`. Even so, through some pre-processing of arguments, the Rust framework does provide support for said var args, and even multi results.
 
 There are X types of var args supported:
-- `OptionalArg<T>` - arguments that can be skipped. Can have multiple per endpoint, but they all must be the last arguments of the endpoint.  
-- `ManagedVarArgs<T>` - can receive any number of arguments. Note that only one `ManagedVarArgs` can be used per endpoint and must be the last argument in the endpoint. Cannot use both `OptionalArg` and `ManagedVarArgs` in the same endpoint.  
-- `ManagedVarArgs<MultiArgN<T1, T2, ... TN>>` - Can be used when you want to receive a variable number of pairs or arguments. For example, let's say you want to receive a variable number of (token ID, nonce) pairs. You would then use `ManagedVarArgs<MultiArg2<TokenIdentifier, u64>>`.  
+- `OptionalArg<T>` - arguments that can be skipped. Can have multiple per endpoint, but they all must be the last arguments of the endpoint.
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:208]: These guys just got renamed:
OptionalValue, ManagedValueEncoded/MultiValueManagedVec, MultiValueN

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 6]
```diff
@@ -198,16 +198,16 @@ Keep in mind that all the mappers can have as many additional arguments for the
 fn list_per_user_pair(&self, first_addr: &ManagedAddress, second_addr: &ManagedAddress) -> VecMapper<T>;
 ```
 
-Using the correct mapper for your situation can greatly decrease gas costs and complexity, so always remember to carefully evaluate your use-case.  
+Using the correct mapper for your situation can greatly decrease gas costs and complexity, so always remember to carefully evaluate your use-case.
 
 ## VarArgs and MultiResults
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:203]: Let's rename to `Variadic inputs and outputs (multi-values)` or some variation on this.
We deprecated these older names.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 7]
```diff
@@ -198,16 +198,16 @@ Keep in mind that all the mappers can have as many additional arguments for the
...
 
-The Rust language does not natively support var args. The best you could probably do in native rust is use `Option<T>`. Even so, through some pre-processing of arguments, the Rust framework does provide support for said var args, and even multi results.  
+The Rust language does not natively support var args. The best you could probably do in native rust is use `Option<T>`. Even so, through some pre-processing of arguments, the Rust framework does provide support for said var args, and even multi results.
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:205]: The Rust language does not natively support variadic arguments. Smart contracts, on the other hand, have no limitations on accepting a variable number of inputs or producing a variable number of results.
To accommodate this, and to make I/O processing succint, the Rust framework provides a number of so-called multi-value types, that deserialize from multiple inputs and serialize to multiple outputs.
Please note that the same types are used both as arguments and as results. This makes sense especially for places like the callbacks, where the results of the asynchronous call become the inputs of the callback.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 8]
```diff
@@ -246,7 +246,7 @@ For the same token ID and skipped nonce, the encodings look like this:
 
 `myOptArgEndpoint@544f4b454e2d313233343536`
 
-As you can see, the argument can be skipped altogether instead of passing a `00` (`None`).  
+As you can see, the argument can be skipped altogether instead of passing a `00` (`None`).
 
 ### ManagedVec\<T\> vs ManagedVarArgs\<T\>
```
Comment (andrei-marinica) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:251]: ManagedVec\<T\> vs ManagedValueEncoded\<T\>

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md (hunk 1 above)
Comment (camilbancioiu) [reply docs/developers/developer-reference/smart-contract-developer-best-practices.md:304]: Replaced with `ManagedBuffer`.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md (hunk 6 above)
Comment (camilbancioiu) [reply docs/developers/developer-reference/smart-contract-developer-best-practices.md:203]: Changed the title of the section.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md (hunk 7 above)
Comment (camilbancioiu) [reply docs/developers/developer-reference/smart-contract-developer-best-practices.md:205]: Replaced the paragraph with yours.

**Review by dorin-iancu — CHANGES_REQUESTED** [review]

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 9]
```diff
@@ -198,27 +198,29 @@ Keep in mind that all the mappers can have as many additional arguments for the
...
+- `MultiValueManagedVec<T1, T2, ... TN>` - Can be used when you want to receive a variable number of pairs or arguments. For example, let's say you want to receive a variable number of (token ID, nonce) pairs. You would then use `MultiValueManagedVec<TokenIdentifier, u64>`.
 
 Note: Keep in mind you have to specify the `#[var_args]` annotation in front of those arguments. For example:
 ```
 #[endpoint(myOptArgEndpoint)]
 fn my_opt_arg_endpoint(&self, obligatory_arg: T1, #[var_args] opt_arg: OptionalArg<T2>) {}
 
 #[endpoint(myVarArgsEndpoint)]
-fn my_var_args_endpoint(&self, obligatory_arg: T1, #[var_args] args: ManagedVarArgs<T2>) {}
+fn my_var_args_endpoint(&self, obligatory_arg: T1, #[var_args] args: ManagedValueEncoded<T2>) {}
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply docs/developers/developer-reference/smart-contract-developer-best-practices.md:220]: Fixed.

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 10]
```diff
@@ -246,9 +249,9 @@ For the same token ID and skipped nonce, the encodings look like this:
 
 `myOptArgEndpoint@544f4b454e2d313233343536`
 
-As you can see, the argument can be skipped altogether instead of passing a `00` (`None`).  
+As you can see, the argument can be skipped altogether instead of passing a `00` (`None`).
 
-### ManagedVec\<T\> vs ManagedVarArgs\<T\>
+### ManagedVec\<T\> vs ManagedValueEncoded\<T\>
```
Comment (dorin-iancu) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:254]:
```suggestion
### ManagedVec\<T\> vs MultiValueEncoded\<T\>
```

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 11]
```diff
@@ -258,72 +261,67 @@ fn my_var_args_endpoint(&self, args: ManagedVec<(TokenIdentifier, u64, BigUint)>
...
 `myVarArgsEndpoint@0000000c_544f4b454e2d313233343536_0000000000000005_00000001_64_0000000c_544f4b454e2d313233343536_000000000000000a_00000002_01f4`
 
 Note: Above, we've separated the parts with `_` for readability purposes only. On the real blockchain, there would be no underscores, everything would be concatenated.
 
-As you can see, that endpoint is very hard to work with. All arguments have to be passed into this big chunk, with nested encoding, which also adds additional lengths for the `TokenIdentifier` (i.e. the 0000000c in front, which is length 12) and for `BigUint` (i.e. the length in bytes).  
+As you can see, that endpoint is very hard to work with. All arguments have to be passed into this big chunk, with nested encoding, which also adds additional lengths for the `TokenIdentifier` (i.e. the 0000000c in front, which is length 12) and for `BigUint` (i.e. the length in bytes).
 
-For the `ManagedVarArgs` approach, this endpoint is a lot easier to use. For the same arguments, the call data looks like this:
+For the `ManagedValueEncoded` approach, this endpoint is a lot easier to use. For the same arguments, the call data looks like this:
```
Comment (dorin-iancu) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:274]:
```suggestion
For the `MultiValueEncoded` approach, this endpoint is a lot easier to use. For the same arguments, the call data looks like this:
```

File: docs/developers/developer-reference/smart-contract-developer-best-practices.md [hunk 12]
```diff
@@ -258,72 +261,67 @@ fn my_var_args_endpoint(&self, args: ManagedVec<(TokenIdentifier, u64, BigUint)>
...
-    // use static_buffer_slice in some way
-}
-```
-
-We recommend not doing this unless you REALLY have to. This code is marked as unsafe for a good reason.  
+| `&[u8]` | `ManagedBuffer` |
+| - | `BigUint` |
+| `Vec<u8>` and `BoxedBytes` | `ManagedBuffer` |
+| `String` | `ManagedBuffer` |
+| `MultiValueVec` | `ManagedValueEncoded` / `MultiValueManagedVec` |
```
Comment (dorin-iancu) [inline docs/developers/developer-reference/smart-contract-developer-best-practices.md:311]:
```suggestion
| `MultiValueVec` | `MultiValueEncoded` / `MultiValueManagedVec` |
```

**Review by dorin-iancu — CHANGES_REQUESTED** [review]

**Review by andrei-marinica — DISMISSED** [review]

**Review by dorin-iancu — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**